### ISSUE TITLE
Custom formatter for enum class - compilation failure with MSVC

### ISSUE THREAD

**Issue by m5k8** [fmtlib/fmt#3255]
The following code:

        #include <fmt/format.h>
        #include <iostream>
        
        enum class my_enum {};
        
        template <> struct fmt::formatter<my_enum>
        {
	        constexpr auto parse(format_parse_context& ctx) -> decltype(ctx.begin())
	        {
		        auto it = ctx.begin(), end = ctx.end();
		        if (it != end && *it != '}') throw format_error("invalid format");
		        return it;
	        }
        
	        template <typename FormatContext>
	        auto format(const my_enum& v, FormatContext& ctx) const -> decltype(ctx.out())
	        {
		        return fmt::format_to(ctx.out(), "{:08X}", (unsigned int)v);
	        }
        };
        
        int main()
        {
	        my_enum x{};
	        auto str = fmt::format("Test {}", x);
	        std::cout << str << std::endl;
  	        return 0;
        }

compiles and works with gcc/clang, but MSVC fails:

    fmt\include\fmt\core.h(1772,7): error C2338: static_assert failed: 'Cannot format an argument. To make type T formattable provide a formatter<T> specialization: https://fmt.dev/latest/api.html#udt'
    fmt\include\fmt\core.h(1791,67): message : see reference to function template instantiation 'fmt::v9::detail::value<Context> fmt::v9::detail::make_value<Context,my_enum&>(T)' being compiled
    1>        with
    1>        [
    1>            Context=fmt::v9::format_context,
    1>            T=my_enum &
    1>        ]
    fmt\include\fmt\core.h(1910,1): message : see reference to function template instantiation 'fmt::v9::detail::value<Context> fmt::v9::detail::make_arg<true,Context,fmt::v9::detail::type::custom_type,my_enum&,0>(T)' being compiled
    1>        with
    1>        [
    1>            Context=fmt::v9::format_context,
    1>            T=my_enum &
    1>        ]
    fmt\include\fmt\core.h(1933,3): message : see reference to function template instantiation 'fmt::v9::format_arg_store<fmt::v9::format_context,my_enum>::format_arg_store<my_enum&>(my_enum &)' being compiled
    fmt\include\fmt\core.h(1933,3): message : see reference to function template instantiation 'fmt::v9::format_arg_store<fmt::v9::format_context,my_enum>::format_arg_store<my_enum&>(my_enum &)' being compiled
    fmt\include\fmt\core.h(3214,20): message : see reference to function template instantiation 'fmt::v9::format_arg_store<fmt::v9::format_context,my_enum> fmt::v9::make_format_args<fmt::v9::format_context,my_enum&>(my_enum &)' being compiled
    test.cpp(29,37): message : see reference to function template instantiation 'std::string fmt::v9::format<my_enum&>(fmt::v9::basic_format_string<char,my_enum &>,my_enum &)' being compiled

When I use some struct{ int v; } instead of enum class for my custom data type, then all is fine.
What am I doing wrong?

**Comment by vitaut** [comment]
I don't see anything obviously wrong in this example. Could you provide a godbolt repro?

**Comment by sukesh-ak** [comment]
@vitaut 

Works in `fmt 8.0.1`
https://www.godbolt.org/z/TMW8xhvvc


Fails with `fmt 9.1.0`
https://www.godbolt.org/z/7eEaeY88d

**Comment by vitaut** [comment]
@sukesh-ak your example is unrelated and the error is expected because enums are not formattable by default. See https://fmt.dev/latest/api.html#formatting-user-defined-types.

**Comment by sukesh-ak** [comment]
> @sukesh-ak your example is unrelated and the error is expected because enums are not formattable by default. See https://fmt.dev/latest/api.html#formatting-user-defined-types.

Then how does it work in `fmt 8.0.1` ?
Have given the link for working and failing.

**Comment by sukesh-ak** [comment]
Formatting of enum to string works like this as per doc. I have tested
https://fmt.dev/latest/api.html

```c++

enum class color {red, green, blue};

template <> struct fmt::formatter<color>: formatter<string_view> {
  // parse is inherited from formatter<string_view>.
  template <typename FormatContext>
  auto format(color c, FormatContext& ctx) const {
    string_view name = "unknown";
    switch (c) {
    case color::red:   name = "red"; break;
    case color::green: name = "green"; break;
    case color::blue:  name = "blue"; break;
    }
    return formatter<string_view>::format(name, ctx);
  }
};

```

**Comment by vitaut** [comment]
In some earlier versions enums happened to "work" because enums were converted to integers. This was fixed in 8.1.

**Comment by vitaut** [comment]
Closing as I wasn't able to reproduce it on https://godbolt.org/z/E6d87h9cc. Feel free to reopen with more repro details.
